Rework key resolution/validation for x5c (tryin' to fix #232) (#239)
oauth-wg/oauth-sd-jwt-vc: draft-ietf-oauth-sd-jwt-vc.md

- JWT VC Issuer Metadata: If a recipient supports JWT VC Issuer Metadata and if the `iss` value contains an HTTPS URI, the recipient MUST
obtain the public key using JWT VC Issuer Metadata as defined in (#jwt-vc-issuer-metadata).

> - X.509 Certificates: If the recipient supports X.509 Certificates, the recipient MUST obtain the public key from the end-entity certificate of the certificates from the `x5c` header parameter of the Issuer-signed JWT and validate X.509 certificate chain accordingly. It MUST also ensure that the `iss` value matches a `uniformResourceIdentifier` SAN entry of the end-entity certificate or that the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.

Reviewer: I find the following part confusing

> or that the domain name in the `iss` value

I believe it would be more clear something like:

It MUST also ensure that the `iss` value matches a `uniformResourceIdentifier` or a `dNSName` SAN entry of the end-entity certificate.

Reviewer: PS: After this PR, my understanding is that the following options would be available

- `iss` is HTTPS URL, `x5c` is not provided, `kid` maybe used (recommended) => Resolve pub key via metadata
- `iss` is DID, `x5c is not provided, `kid` is DID-URL => Resolve pub key via DID resolution
- `iss` is a HTTPS URL, `x5c` is provided  => Pub Key is the leaf certificate, `iss` must be a SAN `uniformResourceIdentifier` entry
- `iss` is a string, `x5c` is provided => Pub key is the leaf certificate, `iss` must be a SAN `dNSName`  entry

Author: Sigh, yeah. I was just going to say something about iss always having to be an HTTPS URL. And then I realized I was confused about what layer what restriction is coming from due to being careless and having carelessly internalized some of the discussion in https://github.com/oauth-wg/oauth-sd-jwt-vc/issues/232 and https://github.com/openid/oid4vc-haip-sd-jwt-vc/issues/103.

Author: ```suggestion
- X.509 Certificates: If the recipient supports X.509 Certificates and the `iss` value contains an HTTPS URI, the recipient MUST obtain the public key from the end-entity certificate of the certificates from the `x5c` header parameter of the Issuer-signed JWT and validate X.509 certificate chain accordingly. It MUST also ensure that the `iss` value matches a `uniformResourceIdentifier` SAN entry of the end-entity certificate or that the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.
```

Reviewer: Dear @bc-pi 
Thanks for the clarification. I think it is clear now.

Author: Thanks for catching the mistake

> - X.509 Certificates: If the recipient supports X.509 Certificates and the `iss` value contains an HTTPS URI, the recipient MUST obtain the public key from the end-entity certificate of the certificates from the `x5c` header parameter of the Issuer-signed JWT and validate X.509 certificate chain accordingly. It MUST also ensure that the `iss` value matches a `uniformResourceIdentifier` SAN entry of the end-entity certificate or that the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.

Reviewer 2: ```suggestion
- X.509 Certificates: If the recipient supports X.509 Certificates and the `iss` value contains an HTTPS URI, the recipient MUST obtain the public key from the end-entity certificate of the certificates from the `x5c` header parameter of the Issuer-signed JWT and validate the X.509 certificate chain accordingly. It MUST also ensure that the `iss` value matches a `uniformResourceIdentifier` SAN entry of the end-entity certificate or that the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.
```

> - X.509 Certificates: If the recipient supports X.509 Certificates and the `iss` value contains an HTTPS URI, the recipient MUST obtain the public key from the end-entity certificate of the certificates from the `x5c` header parameter of the Issuer-signed JWT and validate the X.509 certificate chain accordingly. It MUST also ensure that the `iss` value matches a `uniformResourceIdentifier` SAN entry of the end-entity certificate or that the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.

Reviewer 3: To avoid any ambiguity around the "it" in the second sentence and make it more clear that both steps are important:

```suggestion
- X.509 Certificates: If the recipient supports X.509 Certificates and the `iss` value contains an HTTPS URI, the recipient MUST
     1. obtain the public key from the end-entity certificate of the certificates from the `x5c` header parameter of the Issuer-signed JWT and validate the X.509 certificate chain accordingly, and
     2. ensure that the `iss` value matches a `uniformResourceIdentifier` SAN entry of the end-entity certificate or that the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.
```

(Not sure if the indentation works.)

Author: > (Not sure if the indentation works.)

It seems to have worked okay:
https://drafts.oauth.net/oauth-sd-jwt-vc/bc-iss-not-dns-uri-scheme-wtf-x5t-kid-url-ffs-https-etc-etc/draft-ietf-oauth-sd-jwt-vc.html#name-issuer-signed-jwt-verificat

> - X.509 Certificates: If the recipient supports X.509 Certificates and the `iss` value contains an HTTPS URI, the recipient MUST obtain the public key from the end-entity certificate of the certificates from the `x5c` header parameter of the Issuer-signed JWT and validate the X.509 certificate chain accordingly. It MUST also ensure that the `iss` value matches a `uniformResourceIdentifier` SAN entry of the end-entity certificate or that the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.

Reviewer 3: Is it necessary to say what happens if no `x5c` header is present?

Author: I don't think so - none of this applies then ("If the recipient supports X.509 Certificates and the `iss` value contains an HTTPS URI" = false) and _processing_ falls through to the end of that section,

https://drafts.oauth.net/oauth-sd-jwt-vc/bc-iss-not-dns-uri-scheme-wtf-x5t-kid-url-ffs-https-etc-etc/draft-ietf-oauth-sd-jwt-vc.html#section-3.5-4

> If a recipient cannot validate that the public verification key corresponds to the iss value of the Issuer-signed JWT, the SD-JWT VC MUST be rejected.
